spark/src/main/java/org/apache/iceberg/RewriteManifestsAction.java

```diff
@@ -0,0 +1,490 @@
+
+  private final Encoder<ManifestFile> manifestEncoder = Encoders.javaSerialization(ManifestFile.class);
+  private final Encoder<Entry> entryEncoder = Encoders.javaSerialization(Entry.class);
+  private final Encoder<Bin> binEncoder = Encoders.bean(Bin.class);
+
+  private Predicate<ManifestFile> predicate = manifest -> true;
+  private String stagingLocation = null;
+
+  RewriteManifestsAction(SparkSession spark, Table table) {
+    this.spark = spark;
+    this.sparkContext = new JavaSparkContext(spark.sparkContext());
+    this.table = table;
+    this.specs = table.specs();
+    this.summary = Maps.newHashMap();
+    this.defaultParallelism = Integer.parseInt(
+        spark.conf().get("spark.default.parallelism", "200"));
+    this.snapshotIdInheritanceEnabled = PropertyUtil.propertyAsBoolean(
+        table.properties(),
+        TableProperties.SNAPSHOT_ID_INHERITANCE_ENABLED,
+        TableProperties.SNAPSHOT_ID_INHERITANCE_ENABLED_DEFAULT);
+    this.targetManifestSizeBytes = PropertyUtil.propertyAsLong(
+        table.properties(),
+        TableProperties.MANIFEST_TARGET_SIZE_BYTES,
+        TableProperties.MANIFEST_TARGET_SIZE_BYTES_DEFAULT);
+
+    if (table.io() instanceof HadoopFileIO) {
+      // we need to use Spark's SerializableConfiguration to avoid issues with Kryo serialization
+      SerializableConfiguration conf = new SerializableConfiguration(((HadoopFileIO) table.io()).conf());
+      fileIO = new HadoopFileIO(conf::value);
+    } else {
+      fileIO = table.io();
+    }
+  }
+
+  public RewriteManifestsAction rewriteIf(Predicate<ManifestFile> newPredicate) {
+    this.predicate = newPredicate;
+    return this;
+  }
+
+  public RewriteManifestsAction stagingLocation(String newStagingLocation) {
+    this.stagingLocation = newStagingLocation;
+    return this;
+  }
+
+  @Override
+  public RewriteManifestsAction set(String property, String value) {
+    summary.put(property, value);
+    return this;
+  }
+
+  @Override
+  public RewriteManifestsActionResult execute() {
+    Preconditions.checkArgument(stagingLocation != null, "Staging location must be set");
+
+    List<ManifestFile> matchingManifests = findMatchingManifests();
+    if (matchingManifests.isEmpty()) {
+      return null;
+    }
+
+    Broadcast<FileIO> io = sparkContext.broadcast(fileIO);
+
+    int parallelism = Math.min(matchingManifests.size(), defaultParallelism);
+    JavaRDD<ManifestFile> manifestRDD = sparkContext.parallelize(matchingManifests, parallelism);
+    Dataset<ManifestFile> manifestDS = spark.createDataset(manifestRDD.rdd(), manifestEncoder);
+    Dataset<Entry> manifestEntryDS = manifestDS.flatMap(toEntries(io, specs), entryEncoder);
+
+    try {
+      manifestEntryDS.cache();
+
+      long manifestEntrySizeBytes = computeManifestEntrySizeBytes(matchingManifests);
+      Map<Integer, List<PartitionMetadata>> metadataSizeSummary = computeMetadataSizeSummary(
+          manifestEntryDS,
+          manifestEntrySizeBytes);
```

Comment:
I think I can produce basically the same result as this step with this SQL query:

```
with
  manifests as (select * from db.table.manifests where partition_spec_id = 0),
  entries as (select input_file_name() as manifest, * from db.table.entries where status < 2),
  matching_entry_counts as (
      select count(1) as entry_count, e.data_file.partition
      from entries e join manifests m on m.path = e.manifest
      group by e.data_file.partition)
select * from matching_entry_counts
```

Reply:
Yes, we can definitely compute this part using metadata tables. However, we need to migrate the second part of this action (i.e. writing new manifests) to metadata tables as well. Otherwise, we will read the metadata twice and I think we want to avoid that. We need a typed `Dataset` and actual `DataFile` instances with `StructLike` objects as partitions. Assuming we are processing one partition spec at a time will probably help us as the partition struct will have a fixed type.

Reply:
I mention a couple of challenges in the TODO part of the PR description. I think we can try to map Spark's `Row` into `DataFile` or `StructLike` manually. Any other ideas there, @rdblue?

Reply:
Sorry, this is a big one, so I've been trying to understand the parts of it that require using `MapFunction` and `FlatMapFunction`.

I think that we can use the entries table using a query like the one above to get the data, a global sort to do the bin packing, and then we just need to convert back to `DataFile` to write in a task. For that, I'd build a bean to get the data that implements `DataFile`, and an adapter to `IndexedRecord` that depends on the interface. Here's an incomplete example of the wrapper that can be used to write any `DataFile` implementation to Avro:

```java
  static class IndexedDataFile implements DataFile, IndexedRecord {
    private final Types.StructType partitionType;
    private DataFile wrapped = null;

    IndexedDataFile(Types.StructType partitionType) {
      this.partitionType = partitionType;
    }

    public IndexedDataFile wrap(DataFile bean) {
      this.wrapped = bean;
      return this;
    }

    @Override
    public Object get(int pos) {
      switch (pos) {
        case 0:
          return wrapped.path();
        case 1:
          return wrapped.format();
        case 2:
          return wrapped.partition();
        case 3:
          return wrapped.recordCount();
        ...
      }
      throw new IllegalArgumentException("Unknown field position: " + pos);
    }

    @Override
    public void put(int i, Object v) {
      throw new UnsupportedOperationException("Cannot read into IndexedDataFile");
    }

    @Override
    public Schema getSchema() {
      return AvroSchemaUtil.convert(DataFile.getType(partitionType));
    }

    @Override
    public String path() {
      return wrapped.path();
    }

    @Override
    public FileFormat format() {
      return wrapped.format();
    }

    @Override
    public PartitionData partition() {
      return wrapped.partition();
    }

    @Override
    public long recordCount() {
      return wrapped.recordCount();
    }

    ...
  }
```

I'd do the same wrapping for the `StructLike` partition. Then for each task, you just need to get the bean, wrap it, and write it to your `ManifestsWriter`.

Reply:
Also, we can keep the wrapper class package-private. The writer can wrap every incoming `DataFile`.

Reply:
I added the IndexedDataFile wrapper in #901, so I'll resolve this thread.